### PR TITLE
Fix false negatives for D402

### LINKED ISSUE #284: False negatives for D402
The following function fails the D402 check:

```py
def subclasses(cls):
    """Retrieve all subclasses (direct and transitive) from `cls`.

    :param cls: The class to search.
    :type cls: ``type``

    :returns: A generator of subclasses.
    :rtype: ``iter`` of ``type``
    """
    for subclass in cls.__subclasses__():
        for subsubclass in subclasses(subclass):
            yield subsubclass
        yield subclass
```

Presumably the D402 rule just sees the method name followed by an lparen?

### PR DESCRIPTION
**Pull request by barakbeilin**
I want to resolve #284, I've added the appropriate test.

### CODE REVIEW

File: src/tests/test_cases/test.py [hunk 1]
```diff
@@ -316,8 +316,13 @@ def sdgfsdg23245777():
 
 
 @expect('D402: First line should not be the function\'s "signature"')
-def foobar():
-    """Signature: foobar()."""
+def foobar(foo=1, bar=1):
```
Comment (samj1912) [inline src/tests/test_cases/test.py]: These look good. Can you also add some tests which are false negatives to ensure we *don't* raise errors for them?

**Comment by samj1912** [comment]
Can you also update the release notes with this change please?

File: src/tests/test_cases/test.py (hunk 1 above)
Comment (barakbeilin) [reply src/tests/test_cases/test.py:317]: DONE

File: src/tests/test_cases/test.py [hunk 2]
```diff
@@ -316,10 +314,20 @@ def sdgfsdg23245777():
 
 
 @expect('D402: First line should not be the function\'s "signature"')
-def foobar():
+def foobar(foo=1, bar=1):
+    """Signature: foobar(foo=1, bar=1)."""
+
```
Comment (Nurdok) [inline src/tests/test_cases/test.py:319]: This is actually an interesting case, should we also fail if the docstring contains the keyword arguments without the default value? e.g.,

```py
def foobar(foo=1, bar=1):
   """Signature: foobar(foo, bar)."""
```

Comment (samj1912) [reply src/tests/test_cases/test.py:319]: There can be lots of combinations here with args/kwargs

Comment (barakbeilin) [reply src/tests/test_cases/test.py]: docs.python presents functions with their default values, that's why I decided to check for this kind of signature, other might argue with my logic though. There's no explicit rule about what's a signature of a function in PEP257.

File: src/pydocstyle/checker.py [hunk 3]
```diff
@@ -478,7 +478,10 @@ def check_no_signature(self, function, docstring):  # def context
         """
         if docstring:
             first_line = ast.literal_eval(docstring).strip().split('\n')[0]
-            if function.name + '(' in first_line.replace(' ', ''):
+            function_signarture = function.source.split(
```
Comment (samj1912) [inline src/pydocstyle/checker.py:481]: This might break if the function signature has type hints.

**Comment by samj1912** [comment]
Closing this as it has been inactive for a while. Please feel free to re-open if this is still relevant.

**Pull request closed without merging**